src/scheduler.h

```diff
@@ -38,44 +38,36 @@ class SchedulerTask : public Task
 			return eventId;
 		}
 
-		std::chrono::system_clock::time_point getCycle() const {
-			return expiration;
+		uint32_t getDelay() const {
+			return delay;
 		}
-
 	private:
-		SchedulerTask(uint32_t delay, std::function<void (void)>&& f) : Task(delay, std::move(f)) {}
+		SchedulerTask(uint32_t delay, std::function<void (void)>&& f) : Task(std::move(f)), delay(delay) {}
 
 		uint32_t eventId = 0;
+		uint32_t delay = 0;
 
 		friend SchedulerTask* createSchedulerTask(uint32_t, std::function<void (void)>);
 };
 
 SchedulerTask* createSchedulerTask(uint32_t delay, std::function<void (void)> f);
 
-struct TaskComparator {
-	bool operator()(const SchedulerTask* lhs, const SchedulerTask* rhs) const {
-		return lhs->getCycle() > rhs->getCycle();
-	}
-};
-
 class Scheduler : public ThreadHolder<Scheduler>
 {
 	public:
 		uint32_t addEvent(SchedulerTask* task);
-		bool stopEvent(uint32_t eventId);
+		void stopEvent(uint32_t eventId);
 
 		void shutdown();
 
 		void threadMain();
-
 	private:
 		std::thread thread;
-		std::mutex eventLock;
-		std::condition_variable eventSignal;
-
-		uint32_t lastEventId {0};
-		std::priority_queue<SchedulerTask*, std::deque<SchedulerTask*>, TaskComparator> eventList;
-		std::unordered_set<uint32_t> eventIds;
+		std::atomic<uint32_t> lastEventId {0};
+		std::unordered_map<uint32_t, boost::asio::deadline_timer*> eventIdTimerMap;
+		std::vector<boost::asio::deadline_timer*> timerCacheVec;
```

Comment:
Does it require pointers? I would store instances otherwise. Don't like the sight of two levels of indirections.

Reply:
I believe if i don't use pointers i can't reuse the deadline_timer object. Got ideas? "Don't like the sight of two levels of indirections" elaborate please.

Reply:
Vector is allocated on the heap, and if you use pointers for `deadline_timers` they will also be allocated in the heap, possibly incurring in performance issues if these are accessed  often and out of order. If you use the objects instead of pointers, they are stored contiguouly, so you avoid one round trip to the slow RAM.

Why/when would you reuse the object instead?